### ISSUE TITLE
Source for AI memos

### ISSUE THREAD

**Issue by larsbrinkhoff** [PDP-10/its#948]
MIT publications that we can potentially rebuild from source. (foo) means the source is available in the foo repository.

Document | Source | Type
-- | -- | --
AI Memo 147A | HUR; DDTMEM 188 | TJ6
AI Memo 191 | WP; BIBLIO 191A21
AI Memo 238 | DEE; PRITS 42 | TJ6
AI Memo 239 | MB; HAKMEM 141 | Plain text
[AI Memo 240](http://bitsavers.org/pdf/mit/ai/aim/AIM-240.pdf) | (its) WP; 11SIM 240A38 | TJ6
[AI Memo 260](http://bitsavers.org/pdf/mit/ai/aim/AIM-260.pdf) | (its) WP; LOCK 260A17 | PUB
[AI Memo 261](http://bitsavers.org/pdf/mit/ai/aim/AIM-261.pdf) | (its) WP; PEEK 261A17 | TJ6
[AI Memo 267](http://bitsavers.org/pdf/mit/ai/aim/AIM-267.pdf) | MINSKY; DESIGN 76
[AI Memo 285](http://bitsavers.org/pdf/mit/ai/aim/AIM-285.pdf) | WP; LINFND 285A43 | TJ6
[AI Memo 292](http://bitsavers.org/pdf/mit/ai/aim/AIM-292.pdf) | WP; UT 292A15
[AI Memo 295](http://bitsavers.org/pdf/mit/ai/aim/AIM-295.pdf) | WP; LIGHT 295A13
[AI Memo 300](http://bitsavers.org/pdf/mit/ai/aim/AIM-300.pdf) | WP; ARMDSG 300A1
[AI Memo 301](http://bitsavers.org/pdf/mit/ai/aim/AIM-301.pdf) | WP; ARMCTL 301A1
[AI Memo 307](http://bitsavers.org/pdf/mit/ai/aim/AIM-307.pdf) | LLOGO; MEMO > | Pub
[AI Memo 313](http://bitsavers.org/pdf/mit/ai/aim/AIM-313.pdf) | JEA; LOGMAN > | TJ6
[AI Memo 358](http://bitsavers.org/pdf/mit/ai/aim/AIM-358.pdf) | (its) TJ6; TJ6MEM > | TJ6
[AI Memo 444](http://bitsavers.org/pdf/mit/ai/aim/AIM-444.pdf) | (unlambda) LMDOC; PAPER > | TJ6
[AI Memo 447](http://bitsavers.org/pdf/mit/ai/aim/AIM-447.pdf) | (ToTS) ECC; EDOC > | R
[AI Memo 452](http://bitsavers.org/pdf/mit/ai/aim/AIM-452.pdf) | (its) SCHEME; DOCUME > | TJ6
[AI Memo 453](http://bitsavers.org/pdf/mit/ai/aim/AIM-453.pdf) | (its) SCHEME; META > | TJ6
[AI Memo 495](http://bitsavers.org/pdf/mit/ai/aim/AIM-495.pdf) | (its-vault) BKPH; BRDF > | TJ6
[AI Memo 519](https://archive.org/details/MITAIMemo519) | (tots) RMS; EPAP 49 | TJ6
[AI Memo 519A](http://bitsavers.org/pdf/mit/ai/aim/AIM-519A.pdf) | (tots) RMS; EPAP 4 | Scribe
[AI Memo 528](http://bitsavers.org/pdf/mit/ai/aim/AIM-528.pdf) | (unlambda) LMDOC; CADR > | TJ6
[AI Memo 554](http://bitsavers.org/pdf/mit/ai/aim/AIM-554.pdf) | (emacs) info/emacs.mss | Scribe
[AI Memo 555](http://bitsavers.org/pdf/mit/ai/aim/AIM-555.pdf) | (emacs) info/emacs.mss | Scribe
[AI Memo 569](http://bitsavers.org/pdf/mit/ai/aim/AIM-569a.pdf) | (its-vault) LLOGO; GC DOC |
[AI Memo 585](http://bitsavers.org/pdf/mit/ai/aim/AIM-585.pdf) | ASBELL; R PRIMER | R
[AI Memo 623](http://bitsavers.org/pdf/mit/ai/aim/AIM-623.pdf) | (its-vault) MOON; AMBER > |
[AI Memo 643](http://bitsavers.org/pdf/mit/ai/aim/AIM-643.pdf) | (ToTS) RMS; LEP 59 | Scribe
[AI Memo 644](http://bitsavers.org/pdf/mit/ai/aim/AIM-644.pdf) | (klh10) supdup.mss<br>(tots) RMS; SUPAP 6 | Scribe
AI Memo 694 (unpublished) | (ToTS) SPACY; MEMO 15 | Scribe
[AI WP 16](https://dspace.mit.edu/bitstream/handle/1721.1/41049/AI_WP_016.pdf) | WP; LINES 16W36
AI WP 25 | VIS; VF25 1
AI WP 26 | VIS; VF26 1
AI WP 27 | VIS; VF27 1
AI WP 28 | VIS; VF28 1
AI WP 29 | VIS; VF29 1
AI WP 30 | VIS; VF30 162
AI WP 32 | VIS; VF32 1
AI WP 33 | VIS; VF33 1
[AI WP 52](https://dspace.mit.edu/bitstream/handle/1721.1/41085/AI_WP_052.pdf) | FININ; WP52 2
[AI WP 58](https://dspace.mit.edu/bitstream/handle/1721.1/41092/AI_WP_058.pdf) | WP; FUN-FR 58W1
[AI WP 59](https://dspace.mit.edu/bitstream/handle/1721.1/41094/AI_WP_059.pdf) | WP; GT40 CONTEN<br>(its) .INFO.; URUG MANUAL | TJ6
[AI WP 63](https://dspace.mit.edu/bitstream/handle/1721.1/41098/AI_WP_063.pdf) | WP; HYPOTH 63W1
[AI WP 72](https://dspace.mit.edu/bitstream/handle/1721.1/41107/AI_WP_072.pdf) | WP; XGP 72W1
[AI WP 77](https://dspace.mit.edu/bitstream/handle/1721.1/41112/AI_WP_077.pdf) | DHT; MAPPER INF | TJ6
[AI WP 78](https://dspace.mit.edu/bitstream/handle/1721.1/41113/AI_WP_078.pdf) | (its) WP; FED 78W1 | TJ6
[AI WP 80](https://dspace.mit.edu/bitstream/handle/1721.1/41115/AI_WP_080.pdf) (draft) | (unlambda) LMDOC; CONS > | TJ6
[AI WP 87](https://dspace.mit.edu/bitstream/handle/1721.1/41122/AI_WP_087.pdf) | WP; C.E. 87W1
[AI WP 208](https://dspace.mit.edu/bitstream/handle/1721.1/41040/AI_WP_208A.pdf) | (mit/ai) LMWIND; CHOICE > | Bolio
[AI WP 209](https://dspace.mit.edu/bitstream/handle/1721.1/41042/AI_WP_209.pdf) | (mit/ai) LMWIND; OPERAT > | Bolio
[AI WP 214](https://dspace.mit.edu/bitstream/handle/1721.1/41158/AI_WP_214.pdf) | (its-vault) LLOGO; LMT 22 | Bolio?
AI TR 350? | WP; C.E.TR 350T1
| [CSG memo 144](http://csg.csail.mit.edu/CSGArchives/memos/Memo-144.pdf) | CLU; CLUPAP R | R
SYS.18.01 | (its-archives) SYS.18.01.mss | Scribe
[Dazzle Dart paper](https://dl.acm.org/citation.cfm?doid=1216479.1216482) (preprint) | (its-vault) HAL; DAZZL DART | TJ6
[Numerical Shape paper](https://www.sciencedirect.com/science/article/pii/0004370281900230) (preprint) | (its-vault) BKPH; SEM > | TJ6
Lisp Machine Manual | (unlambda) LMMAN; | Bolio
[MTB-418](http://multicians.org/mtbs/mtb418.pdf) | DLW2; BSGROE/BSGFNE
Mathlab Memo #3 "Introduction to ITS for the MACSYMA User" | TMQ; MC > | PUB
Mathlab Memo #3 "Introduction to ITS for the MACSYMA User" | TMQ; MCSRC > | Scribe

Many of these will have missing diagrams that were added after printing.

Some attempts at this: https://stuff.offog.org/its/remastered/

**Comment by atsampson** [comment]
I've added more of these to the list above.

**Comment by larsbrinkhoff** [comment]
I added SYS.18.01 "A Machine-Independent MDL" which is a Scribe document.

**Comment by larsbrinkhoff** [comment]
SAIL; FAIL PUB, but now straying outside MIT.

**Comment by larsbrinkhoff** [comment]
Adding: AI memos 240, 260, 261, 292.  Woking papers 78, 72.  Technical report 350?

**Comment by larsbrinkhoff** [comment]
Some working papers are called "Vision Flash" and reside in VIS; VF* *.

**Comment by atsampson** [comment]
Added AIM 307, the source location of which is mentioned in LLOGO RECENT.

**Comment by atsampson** [comment]
Added AIM 554/555, the Emacs manual, the source for which is on the TOPS-20 Emacs tape. The Emacs info file is also generated from the same source (i.e. it's a proto-Texinfo setup).

**Comment by eswenson1** [comment]
Would it make sense if I committed the binary for PUB, PUB2, and SCRIBE while trying to get the right pieces to build these from source?

**Comment by larsbrinkhoff** [comment]
Sure, binaries are ok when there's a program we want to use and it's hard or impossible to build from source.  For example, I think it's ok to have F40 in only binary form.  We had a FAIL binary for a while until I found the source on Saildart.

**Comment by atsampson** [comment]
Added Multics Technical Bulletin 418 as another potential one - see [the list of MTBs](http://multicians.org/mtb-index.html).

**Comment by eswenson1** [comment]
By the way if any of you guys want an account on a publicly-available Multics system, let me know.

**Comment by larsbrinkhoff** [comment]
#1161 adds four memos to this repository.

**Comment by larsbrinkhoff** [comment]
AI WP 214 - LLOGO; LMTDOC 22.

**Comment by larsbrinkhoff** [comment]
AI memo 623 - MOON; AMBER >

**Comment by larsbrinkhoff** [comment]
AI memo 569 - LLOGO; GC DOC

**Comment by larsbrinkhoff** [comment]
Trying to find the source of HAKMEM - AI memo 239.  The memo says it's HAKMEM 140.  I see these in backup listings:

```
MB     HAKMEM 17         2      0/0/0 00:00:00
MB     HAKMEM 141        1      3/16/0 00:23:10.5
RCC    HAKMEM HAKMEM     1      2/9/0 23:10:13.5
ES     HAKMEM 25         1      2/27/1 21:11:10
ES     HAKMEM 26         1      3/1/1 20:54:23
ES     HAKMEM 30         5      9/24/73 19:15:57.5
ROWE   HAKMEM 29         6      11/15/75 23:20:09.5
ROWE   HAKMEM 57         3      11/22/75 22:43:15
ROWE   HAKMEM 58         3      12/17/75 00:40:09
ROWE   HAKMEM XGP        2      11/22/75 22:40:12.5
ROWE   HAKMEM 29         4      11/15/75 23:20:09.5
ROWE   HAKMEM 59         4      12/18/75 21:44:04
```

I have contacted the people I believe to be MB, ES, and ROWE.

**Comment by larsbrinkhoff** [comment]
- [x] ES doesn't have it.
- [x] ROWE ...
- [ ] MB
- [ ] RCC

**Comment by larsbrinkhoff** [comment]
MB wrote me:

>  I don’t have any file of HAKMEM.  I cannot remember for absolute certain, but I am 99 percent sure that it was not in TJ6 or any text processing format.  I believe it was a straight ASCII file.  I think I consciously did it that way because I wanted to force a lot of formatting myself, manually.  For instance, there are several places that have “ASCII art” diagrams of things.  And I preferred to control line breaks and page breaks, to help convey the continuity or grouping of ideas.  There are some illustrations that are hand drawn (such as the high performance video deflection amplifier schematic), some pen and ink plotter drawn, and some photocopied (such as the page from Knuth).  All this was easier to manipulate when I had total control of the spacing.

**Comment by larsbrinkhoff** [comment]
LMDOC; OPERAT > is AI WP 209.

LMDOC; PROGR > is AI memo 444.

**Comment by ams** [comment]
I think you mean LMWIND; OPERAT >.

**Comment by ams** [comment]
AI memo 444 is LMDOC; PAPER >

**Comment by larsbrinkhoff** [comment]
Thanks, yes it's LMWIND.  444 is already listed in the first comment.  But which memo is PROGR?

**Comment by ams** [comment]
I don't know, I tried poking about and I recall reading this in a AIM.

**Comment by larsbrinkhoff** [comment]
AIM 238 is DEE; PRITS 42.

**Comment by larsbrinkhoff** [comment]
DEE; JDOCU 2 says the title is "ITS 2.0 Reference Manual" from May 1971.  It seems to be written in the JUST text justifier language, DEE; JUST 114.

DEE; ITS WR ITE UP contains both "ITS 2.9 Reference Manual" and "ITS 3.2 Reference Manual".

**Comment by larsbrinkhoff** [comment]
Adding AIM 147A, which is close to HUR; DDTMEM 188 dated 1971-09-10.

@ericosman, is this your document?

**Comment by larsbrinkhoff** [comment]
Found MB; HAKMEM 141

**Comment by ericosman** [comment]
Yes, sounds like mine.  My login for ITS was HUR.  I documented DDT while I was working at the MIT AI lab during the summers from 1970 through 1974.  I worked for Donald Eastlake who worked for Pat Winston.  People I worked with:  Ron Lebel, Mike Beeler, Richard Greenblatt, Bill Gosper, Richard Stallman,

**Comment by larsbrinkhoff** [comment]
Added the unpublished AI memo 694 "ITS internals" by @cstacy: SPACY; MEMO 15.

```
At the time of this writing, ITS has been in use for over a decade.
Until recently, most of the research at the MIT AI Lab was conducted
on the original ITS computer.  With the advent of such computers as
Lisp Machines, and the beginnings of radical devices like Connection
Machines, PDP-10 computers and their operating systems are becoming
obsolete tools.


This document is intended for the last people who will be maintaining
the ITS computers, and for historical purposes.  It might also be of
interest to the designers of future operating systems.

**Comment by larsbrinkhoff** [comment]
Added AI WP 52 "Tracking Wires on Printed Circuit Boards" by @finin, source code FININ; WP 2 timestamped 1973-12-06.

**Comment by larsbrinkhoff** [comment]
DHT; MAPPER INF is AI WP 77.

**Comment by larsbrinkhoff** [comment]
ECC; EDOC > is AI memo 447.

**Comment by larsbrinkhoff** [comment]
ASBELL; R PRIMER is AI memo 585.  
CLU; CLUPAP R is CSG memo 144.

**Comment by larsbrinkhoff** [comment]
Added second file for 644: RMS; SUPAP 6.  643: LEP 59.  519 and 519A: apparently it was converted from TJ6 (RMS; EPAP 49) to Scribe (RMS; EPAP 4).

**Comment by larsbrinkhoff** [comment]
How to tell various formatter source code apart?  Many use the RUNOFF syntax where a period first on a line is a command.

- If it says `.xgp`, it's TJ6.
- If it says `.device xgp` or some other device, it's PUB.
- R sources probably include R MACROS somewhere.

**Comment by larsbrinkhoff** [comment]
AI memo 519 "*EMACS &mdash; The Extensible, Customizable, Self-Documenting Display Editor*"

The first revision of the memo is written in TJ6.  The second, 519A, was converted to Scribe.

Scan by EricIO: https://archive.org/details/MITAIMemo519

New rendition by XGP emulator from RMS;EPAP 49.  [epap.pdf](https://github.com/PDP-10/its/files/14220541/epap.pdf)

**Comment by eswenson1** [comment]
The newly created XGP output has a 2022 date. I assume that is when you converted it to XGP format using TJ6 or Scribe?  (Which formatted did you use?)

**Comment by qu1j0t3** [comment]
@larsbrinkhoff Looks good! I wonder if METAFONT parameters for device compensation (spread/thickening e.g.) were ever published for XGP; that might give a tiny hint as to device characteristics?

**Comment by larsbrinkhoff** [comment]
@eswenson1, it's the timestamp from the XGP file generated by TJ6.  For some reason the ITS clock is way off!  Or else it would have been February 2024.

I didn't convert to Scribe.  The original, 519, was written in TJ6.  It was converted (possibly also revised) in the 80s, resulting in AI memo 519A.

**Comment by larsbrinkhoff** [comment]
@qu1j0t3, intersting!  I didn't even know that was a thing.  What *exactly* is spread/thickening?

**Comment by qu1j0t3** [comment]
@larsbrinkhoff A notable feature of METAFONT is parametric adjustment of bitmaps to suit device characteristics. Every printing device has different amounts of "spread" through mechanical, chemical and optical processes. Without compensation, fonts can render too light or too dark. In fact a common complaint about Computer Modern Roman itself is that uncompensated outlines or bitmaps render "spindly" and faint on modern, precise devices like high resolution screens or offset lithography. The XGP undoubtedly has its unique characteristics here and since it is coëval with METAFONT usage, one might think that approximate METAFONT parameters for this were tested and published. 

Quickly found reference to device parameters, for example: https://tug.org/TUGboat/tb08-2/tb18beet.pdf ; or see the METAFONTbook and CMR source files for full details.

**Comment by qu1j0t3** [comment]
All that said, your pdf looks pretty convincing to me :-)

Is it possible to get original XGP printouts from the personal archives of MIT people?

**Comment by larsbrinkhoff** [comment]
Found "Mathlab Memo #3" *An Introduction to ITS for the MACSYMA User".  TMQ; MC > is TJ6 source, TMQ; MCSRC > is converted to Scribe.

**Comment by eswenson1** [comment]
When I try to convert that with TJ6, I get:
```
:tj6 ejs1;mc >
!XGPCO  unknown command  DSK:EJS1;MC 260(1,0)(0,1)
 DSK:EJS1;MC 260(1,0)(0,1)
.VAL 0; 7414>>JRST 16147
```

**Comment by larsbrinkhoff** [comment]
Right, it's looks a bit strange.  I thought from the ^F font commands it was TJ6, but then maybe not.  I don't know what it is!

**Comment by larsbrinkhoff** [comment]
I found this in another file, which looks much like the MC file.  I believe this is PUB.

```
.device xgp
.!xgpcommand_";squish"
.font 1 "25vg kst"
.turn on "^F" for "%"
.margins(1000,1000,1000,1000)
```

**Comment by eswenson1** [comment]
Trying PUB:
```
ES>pub↑K!
*ejs1;mc >
INVALID INDEX NO. 1 FOR ARRAY MONTH
CALLED FROM 500134  LAST SAIL CALL AT 511104
?
```

**Comment by larsbrinkhoff** [comment]
@cstacy's AI memo 235 "*Getting Started Computing at the AI Lab*" is CSTACY;AIHI MSS; a Scribe document.

(We are not sure about how to get Scribe running on ITS, see #842)

**Comment by larsbrinkhoff** [comment]
Dug up AI memos 353 and 379, the first of the "Lambda the Ultimate" series.

**Comment by larsbrinkhoff** [comment]
Some memos are now reprinted here:  
https://github.com/larsbrinkhoff/ai-memo-reprints
